### PR TITLE
Wire up the rate limiter to net/rpc calls

### PR DESCRIPTION
**Pull request by analogue**
### Description
net/rpc calls to the consul server are now subject to rate limiting.

### Testing & Reproduction steps
- Skipped adding unit tests as there are no newly added conditionals in this PR. Forthcoming integration tests in a separate PR will exercise high level rate limiting use cases.
- Verified manually that net/rpc calls are intercepted and evaluated by the rate limiter.

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: agent/rpc/middleware/interceptors.go [hunk 1]
```diff
@@ -157,3 +159,19 @@ func GetNetRPCInterceptor(recorder *RequestRecorder) rpc.ServerServiceCallInterc
...
+		op := rpcRate.Operation{
+			Name:       reqServiceMethod,
+			SourceAddr: sourceAddr,
+			Type:       rpcRateLimitSpecs[reqServiceMethod],
+		}
+
+		// net/rpc does not provide a way to encode the nuances of the
+		// error response (retry or retry elsewhere) so the error string
+		// from the rate limiter is all that we have.
+		return requestLimitsHandler.Allow(op)
```
Comment (dhiaayachi) [inline agent/rpc/middleware/interceptors.go:175]: Not specific to this PR but we need to make sure that the error returned here have a string that is unique enough to be able to reason about on the client side. As for net/RPC (unlike gRPC) the string error is returned as is in the wire and there is no concepts of error codes and retryable errors.So we need to do the check on the client side based on the string.

Nothing actionable in my comment, just some thinking 😅

**Review by dhiaayachi — COMMENTED** [review]
Thank you for picking this up @analogue!

I added some thoughts and a comment.

File: agent/consul/server.go [hunk 2]
```diff
@@ -469,14 +469,17 @@ func NewServer(config *Config, flat Deps, externalGRPCServer *grpc.Server, incom
 		incomingRPCLimiter:      incomingRPCLimiter,
 	}
 
+	handler, ok := incomingRPCLimiter.(*rpcRate.Handler)
+	if ok {
+		handler.Register(s)
+	}
```
[the remark this answers is not in the document]
Comment (analogue) [reply agent/consul/server.go]: Updated in https://github.com/hashicorp/consul/pull/15879/commits/61f20fdef64d059664ba638097063ad7e3cbeabd

File: agent/rpc/middleware/interceptors.go (hunk 1 above)
Comment (analogue) [reply agent/rpc/middleware/interceptors.go:175]: Yup, the error string is one of these:
```Go 
	ErrRetryElsewhere = errors.New("rate limit exceeded, try a different server")
	ErrRetryLater = errors.New("rate limit exceeded, try again later")
```

**Comment by analogue** [comment]
@boxofrad @dhiaayachi Description updated and ready for another 👀

**Review by boxofrad — APPROVED** [review]
Looking great, nice job! 👏🏻

**Review by jmurret — APPROVED** [review]
Nice work @analogue !

File: agent/rpc/middleware/interceptors.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (analogue) [reply agent/rpc/middleware/interceptors.go:175]: Thanks! Ticketed as follow up work.

**Pull request merged**